core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -848,6 +908,60 @@ private[spark] object JsonProtocol {
     new ExecutorInfo(executorHost, totalCores, logUrls)
   }
 
+  def blockUpdatedInfoFromJson(json: JValue): BlockUpdatedInfo = {
+    BlockUpdatedInfo(
+      blockManagerIdFromJson(json \ "Block Manager ID"),
+      blockIdFromJson(json \ "Block ID"),
+      storageLevelFromJson(json \ "Storage Level"),
+      (json \ "Memory Size").extract[Long],
+      (json \ "Disk Size").extract[Long],
+      (json \ "ExternalBlockStore Size").extract[Long]
+    )
+  }
+
+  def blockIdFromJson(json: JValue): BlockId = {
+    val rddBlockId = Utils.getFormattedClassName(RDDBlockId)
+    val shuffleBlockId = Utils.getFormattedClassName(ShuffleBlockId)
+    val shuffleDataBlockId = Utils.getFormattedClassName(ShuffleDataBlockId)
+    val shuffleIndexBlockId = Utils.getFormattedClassName(ShuffleIndexBlockId)
+    val broadcastBlockId = Utils.getFormattedClassName(BroadcastBlockId)
+    val taskResultBlockId = Utils.getFormattedClassName(TaskResultBlockId)
+    val streamBlockId = Utils.getFormattedClassName(StreamBlockId)
+
+    (json \ "Class Name").extract[String] match {
```

Comment:
It looks like ReplayListenerBus ties to log exceptions except when it receives an IOException (which might be fatal from its perspective).  AFAIK it will try to keep going when encountering unknown events, but I don't know that there's a test for this. I'll investigate.